pernosco-submit

```diff
@@ -796,40 +797,65 @@ def upload(pernosco_user, pernosco_group, pernosco_user_secret_key):
 
-def get_crypto_data_for_package(package_file_name):
+def extract_crypto_data_from_package(package_file_name):
     """
-    Return the persisted cryptographic data required to upload the specified package file.
+    Extract the persisted cryptographic data required to upload the specified package file
+    from said package file.
+    NOTE: This alters the package file by removing the extracted cryptographic data.
 
         Parameters:
             package_file_name (str): Name of the package file to be uploaded to Pernosco
 
         Returns:
-            A tuple of cryptographic data required for uploading specified package: (public_key, signature, nonce)
+            crypto_data (dict): Cryptographic data required for uploading specified package
     """
-    public_key = None
-    signature = None
-    nonce = None
-    with open("%s.crypto"%package_file_name, "r") as crypto_file:
-        public_key = ""
-        is_reading_public_key = True
-        while is_reading_public_key:
-            current_line = crypto_file.readline()
-            public_key += current_line
-            if "END PUBLIC KEY" in current_line:
-                is_reading_public_key = False
-
-        # Skip empty newline.
-        crypto_file.readline()
-
-        signature = crypto_file.readline()
-
-        # Skip empty newline.
-        crypto_file.readline()
-
-        nonce = crypto_file.readline()
-    return (public_key, signature, nonce)
+    print("Extracting cryptographic data required for upload from %s"%package_file_name)
+
+    # First, get the length of the cryptographic data JSON string from the end
+    # of the file.
+    crypto_file = open(package_file_name, "r+")
+    crypto_file.seek(0, io.SEEK_END)
+
+    is_reading_json_len = True
+    offset = crypto_file.tell() - 1
+    json_len_str = ""
+    while is_reading_json_len:
+        crypto_file.seek(offset, io.SEEK_SET)
+        byte = crypto_file.read(1)
+        if byte == "\n":
+            is_reading_json_len = False
+        else:
+            json_len_str += byte
+            offset -= 1
+
+    # Reverse the length string since it was read backwards from the end of the fle.
+    json_len_str= json_len_str[::-1]
+    json_len = int(json_len_str)
+
+    # Second, truncate the file so that the JSON string length is removed.
+    crypto_file.seek(offset, io.SEEK_SET)
+    crypto_file.truncate()
+    crypto_file.close()
+
+    # Third, parse the cryptographic data JSON string from the end of the file.
+    crypto_file = open(package_file_name, "r+")
+    crypto_file.seek(0, io.SEEK_END)
+    crypto_file.seek(crypto_file.tell() - json_len, io.SEEK_SET)
+    crypto_data = json.load(crypto_file)
+
+    # Finally, truncate the file so that the cryptographic data JSON string is
+    # removed.
+    crypto_file.seek(0, io.SEEK_END)
+    crypto_file.seek(crypto_file.tell() - json_len, io.SEEK_SET)
+    crypto_file.truncate()
+    crypto_file.close()
```

Comment:
This can be simplified a bit by passing the size parameter to `truncate()`. Also I think we only need one final truncate; you should be able to seek to the start of the JSON, read one line to get the JSON, then truncate the file to the correct size.

Reply:
Done - thanks for the tip!